src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -1448,14 +1446,25 @@ Klass* InstanceKlass::array_klass_impl(bool or_null, int n, TRAPS) {
   }
   // _this will always be set at this point
   ObjArrayKlass* oak = array_klasses();
-  if (or_null) {
-    return oak->array_klass_or_null(n);
-  }
   return oak->array_klass(n, THREAD);
 }
 
-Klass* InstanceKlass::array_klass_impl(bool or_null, TRAPS) {
-  return array_klass_impl(or_null, 1, THREAD);
+Klass* InstanceKlass::array_klass_or_null(int n) {
+  // Need load-acquire for lock-free read
+  if (array_klasses_acquire() == NULL) {
+    return NULL;
+  }
+  // _this will always be set at this point
```

Comment:
I have a hard time understanding the meaning of the above comment. Maybe it should be removed? (Same foe line 1447).

Reply:
Until you quoted it I hadn't noticed the underscore on _this. I read it as "this [array_klasses()] will always be set [non-NULL] at this point".

I can restructure this as an if-else to make it crystal clear, and avoid the re-read of array_klasses().